core/domain/stats_services.py

```diff
@@ -18,22 +18,155 @@
 
 import itertools
 
-from core.domain import exp_services
+from core.domain import exp_domain
 from core.domain import interaction_registry
 from core.domain import stats_domain
-from core.domain import stats_jobs_continuous
 from core.platform import models
 
 (stats_models,) = models.Registry.import_models([models.NAMES.statistics])
 
+# Counts contributions from all versions.
+VERSION_ALL = 'all'
+
+def handle_stats_creation_for_new_exploration(exploration):
+    """Creates ExplorationStatsModel for the freshly created exploration and
+    sets all initial values to zero.
+
+    Args:
+        exploration: Exploration. The exploration domain object.
+    """
+    state_stats_mapping = {
+        state_name: stats_domain.StateStats.create_default()
+        for state_name in exploration.states
+    }
+
+    exploration_stats = stats_domain.ExplorationStats(
+        exploration.id, exploration.version, 0, 0, state_stats_mapping)
+    create_stats_model(exploration_stats)
+
+
+def handle_stats_creation_for_new_exp_version(exploration, change_list):
+    """Retrieves the ExplorationStatsModel for the old exp_version and makes
+    any required changes to the structure of the model. Then, a new
+    ExplorationStatsModel is created for the new exp_version.
+
+    Args:
+        exploration: Exploration. The exploration domain object after the
+            commit.
+        change_list: list(dict). A list of changes introduced in this commit.
+    """
+    old_exp_version = exploration.version - 1
+    new_exp_version = exploration.version
+    exploration_stats = get_exploration_stats_by_id(
+        exploration.id, old_exp_version)
+    if exploration_stats is None:
+        handle_stats_creation_for_new_exploration(exploration)
+        return
+
+    # This mapping from new state names to old ones account for circular
+    # renames and multiple renames within a commit. We will use this mapping
+    # to handle state renames in the statistics model.
+    new_to_old_state_names = exploration.get_state_names_mapping(change_list)
+
+    # Handling state renames.
+    for state_name in exploration.states:
+        if state_name not in new_to_old_state_names.values():
```

Comment:
Compute old_state_names = new_to_old_state_names.values() initially, outside the loop, so that you're not repeating the same computation N times.